### PR TITLE
Migrating StackTracesSpec

### PR DESCRIPTION
**Pull request by serragnoli**

### CODE REVIEW

**Comment by adamgfraser** [comment]
This is looking good! I know this can be a tricky one because the traces are supposed to have particular strings in a particular order but it looks like you have solved the biggest problem of getting the tracing information reporting with ZIO Test. Is there anything you are struggling with or could use help on?

**Comment by serragnoli** [comment]
Thanks Adam.
Actually I got stuck in the next test, some error I can't remember now.
I was going to paste the snippet here now but my stashed changes conflicted badly with the version I committed... getting it fixed now 😄

**Comment by serragnoli** [comment]
Thanks Adam.
Actually I got stuck in the next test, some error I can't remember.
I was going to paste the snippet here but my stashed changes conflicted badly with the version I committed... getting it fixed now 😄

**Comment by serragnoli** [comment]
Thanks Adam.
I got stuck on `foreachPar`.

```bash
Warning: A test is using time, but is not advancing the test clock, which may result in the test hanging. Use TestClock.adjust to manually advance the time.
[info] - StackTracesSpec
[info]   + basic test
[info]   + foreachTrace
[info]   + foreach fail
[info]   - foreachPar fail
[info] Timeout of 1 m exceeded.
[info] Ran 4 tests in 1 m: 3 succeeded, 0 ignored, 1 failed
[info] ZIO Test
[info] - StackTracesSpec
[info]   - foreachPar fail
[info] Timeout of 1 m exceeded.
[info] Done
[error] Failed: Total 4, Failed 1, Errors 0, Passed 3
[error] Failed tests:
[error] 	zio.StackTracesSpec_ToZioMigration
[error] (coreTestsJVM / Test / testOnly) sbt.TestsFailedException: Tests unsuccessful
```

Code that generated the above logs
```scala
object StackTracesSpecUtil {
// [...]

  def foreachParFail: ZIO[Clock, Nothing, Unit] =
    for {
      _ <- ZIO.foreachPar(1 to 10) { i =>
            ZIO.sleep(1.second) *> (if (i >= 7) UIO(i / 0) else UIO(i / 10))
        }
    } yield ()

  implicit final class CauseMust[R >: ZEnv](io: ZIO[R, Any, Any]) {
    def causeMust(check: Cause[Any] => TestResult) = {
      io.foldCause[TestResult](
        cause => {
          show(cause)
          check(cause)
        },
        _ => assert(true, isFalse)
      )
    }
  }

object StackTracesSpec_ToZioMigration extends ZIOBaseSpec(
// [...]

        testM("foreachPar fail") {
          import StackTracesSpecUtil._

          val io = for {
            trace <- StackTracesSpecUtil.foreachParFail
          } yield trace

          io causeMust { cause =>
            assert(cause.traces.head.stackTrace.size, equalTo(2)) &&
              assert(cause.traces.head.stackTrace.exists {
                (_: ZTraceElement) match {
                  case s: SourceLocation => s.method contains "foreachParFail"
                  case _ => false
                }
              }, isTrue)
          }
        }
      )
    )
```

**Comment by adamgfraser** [comment]
Okay, well that is a problem that has an easy fix and our warning messages about it are working so that is great! Effects involving time in ZIO Test by default use the `TestClock`, which requires you to manually adjust the clock using `TestClock.adjust` or `TestClock.setTime`. So I would look at either using `TestClock.setTime(1.second)` before the `foreachPar` call or consider deleting the `sleep` call entirely if possible.

**Comment by serragnoli** [comment]
```scala
ZIO.foreachParN(4)(1 to 10) { i =>  ZIO.sleep(1 second) *> (if (i >= 7) UIO(i / 0) else UIO(i / 10))
```
The lambda above is not being executed due to `ZIO.sleep(1 second)` which is causing the test to timeout in ZIO. Dropping the `sleep` makes the test succeed.

Is this the expected behaviour?

**Comment by adamgfraser** [comment]
If you are using `sleep` there and not adjusting the `TestClock` the test is definitely going to time out and that is expected. So the question is can we just eliminate the `sleep` call or do we need to keep it and add a `TestClock.adjust` call. It doesn't seem strictly necessary to me since we aren't checking that the `sleep` call is included in the trace, though maybe the point is to make sure that tracing works for effects that are really asynchronous. To stay true to the original test I would try something like this:

```scala
  def foreachParNFail = {
    val io = for {
      fiber <- ZIO.foreachParN(4)(1 to 10) { i =>
            ZIO.sleep(1.second) *> (if (i >= 7) UIO(i / 0) else UIO(i / 10))
          }.fork
          _ <- TestClock.adjust(1.second)
      _ <- fiber.join
    } yield ()
```

That should be the same, though I don't know if it will result in a slightly different trace so we may have to update the expected result accordingly.

File: core-tests/jvm/src/test/scala-2.12/zio/StackTracesSpec.scala [hunk 1]
```diff
@@ -0,0 +1,658 @@
+package zio
```
[the remark this answers is not in the document]
Comment (neko-kai) [reply core-tests/jvm/src/test/scala-2.12/zio/StackTracesSpec.scala]: Traces aren't supported on Scala.js though, I guess no point running this test there.

[the remark this answers is not in the document]
Comment (adamgfraser) [reply core-tests/jvm/src/test/scala-2.12/zio/StackTracesSpec.scala]: Ah, I hadn't realized that. Yes we should definitely keep it where it is then. Sorry about that!

**Comment by ghostdogpr** [comment]
Since that file is the last one to be migrated, maybe we can remove the dependency to specs2?

**Comment by adamgfraser** [comment]
Can you also merge upstream changes? We just changed the way that specs are written to be:

```scala
object StackTracesSpec extends DefaultRunnableSpec {
  def spec = suite("StackTracesSpec)(...)
}
```

**Comment by serragnoli** [comment]
I will take care of all the pending comments and other bits of this PR in couple of hours

**Comment by adamgfraser** [comment]
Great! You should also be able to get rid of the separate `Util` object with the new inheritance style of writing specs.

File: core-tests/jvm/src/test/scala-2.12/zio/StackTracesSpec.scala [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+          import StackTracesSpecUtil._
+
+          StackTracesSpecUtil.fiberAncestryIsLimitedFixture.recursiveFork(10000) causeMust { cause =>
+            assert(cause.traces.size, equalTo(1)) &&
+            assert(cause.traces.head.parents.size, equalTo(10)) &&
+            assert(cause.traces.head.executionTrace.exists(_.prettyPrint.contains("recursiveFork")), isTrue) &&
+            assert(cause.traces.head.parents.head.stackTrace.exists(_.prettyPrint.contains("recursiveFork")), isTrue)
+          }
+        },
+        testM("blocking trace") {
```
[the remark this answers is not in the document]
Comment (serragnoli) [reply core-tests/jvm/src/test/scala-2.12/zio/StackTracesSpec.scala]: Is this still applicable in light of this [comment](https://github.com/zio/zio/pull/2247#discussion_r347176364)?

[the remark this answers is not in the document]
Comment (adamgfraser) [reply core-tests/jvm/src/test/scala-2.12/zio/StackTracesSpec.scala]: You can disregard this one since we will be keeping the whole thing JVM only.

**Comment by serragnoli** [comment]
> Great! You should also be able to get rid of the separate `Util` object with the new inheritance style of writing specs.

Do you mean from
```scala
testM("basic test") {
      for {
        trace <- StackTracesSpecUtil.basicTest
      } yield {
        StackTracesSpecUtil.show(trace)
        assert(trace.executionTrace.size, equalTo(4)) &&
```
to 
```scala
testM("basic test") {
      for {
          _     <- ZIO.unit
          trace <- ZIO.trace
      } yield {
        StackTracesSpecUtil.show(trace)

        assert(trace.executionTrace.size, equalTo(4)) &&
```

**Comment by adamgfraser** [comment]
I just meant with the new style you can do:

```scala
object StackTracesSpec extends DefaultRunnableSpec {
  def spec = suite("StackTracesSpec")(...)

  // helper methods go here
}
```

So you don't need to have a separate object for your helpers like now.

**Pull request merged**